### PR TITLE
Hash the URI as part of the hyperlink ID

### LINKED ISSUE #7698: Hyperlinks with the same ID and different URIs collide when they shouldn't
In Windows Terminal Preview 1.4.2652.0.
Hyperlinks support is a great addition, but from my understanding there is a bug with how links IDs are parsed and handled.

> Character cells that have the same target URI and the same nonempty id are always underlined together on mouseover.
> The same id is only used for connecting character cells whose URIs is also the same. Character cells pointing to different URIs should never be underlined together when hovering over.
(Source: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda#hover-underlining-and-the-id-parameter )

But testing the same ID with different URIs shows Windows Terminal is grouping them together and using the URI of the last one for both :
`echo -e '\e]8;id=a;http://www.microsoft.com/\aMicrosoft\e]8;;\a \e]8;id=a;http://www.google.com/\aGoogle\e]8;;\a'`

image

Even different IDs are grouped together if they are not numbers:
`echo -e '\e]8;id=a;http://www.microsoft.com/\aMicrosoft\e]8;;\a \e]8;id=b;http://www.google.com/\aGoogle\e]8;;\a'`

While the page about hyperlinks explicitely shows and states that these IDs are strings, not numbers:

> params is an optional list of key=value assignments, separated by the : character. Example: id=xyz123:foo=bar:baz=quux.

> Complex apps that display data that might itself contain OSC 8 hyperlinks (such as terminal multiplexers, less -R) should do the following: If the encountered OSC 8 hyperlink already has an id, they should prefix it with some static string, or if multiple windows/panes are supported by the app, a prefix that's unique to that window/pane to prevent conflict with other windows/panes.

### PR DESCRIPTION
**Pull request by DHowett**
It turns out that we missed part of the OSC 8 spec which indicated that
_hyperlinks with the same ID but different URIs are logically distinct._

> Character cells that have the same target URI and the same nonempty id
> are always underlined together on mouseover.
> The same id is only used for connecting character cells whose URIs is
> also the same. Character cells pointing to different URIs should never
> be underlined together when hovering over.

This pull request fixes that oversight by appending the (hashed) URI to
the generated ID.

When Terminal receives one of these links over ConPTY, it will hash the
URL a second time and therefore append a second hashed ID. This is taken
as an acceptable cost.

Fixes #7698

### CODE REVIEW

**Comment by DHowett** [comment]
GOSH, it's like I didn't even compile the tests.

**Review by PankajBhojwani — APPROVED** [review]

**Review by miniksa — COMMENTED** [review]
Looks fine, but waiting to see the test changes.

**Comment by DHowett** [comment]
OKAY I ADDED A NEW TEST GOSH

File: src/buffer/out/textBuffer.cpp [hunk 1]
```diff
@@ -2279,32 +2279,35 @@ std::wstring TextBuffer::GetHyperlinkUriFromId(uint16_t id) const
...
+        numericId = _currentHyperlinkId;
         ++_currentHyperlinkId;
     }
     else
     {
         // assign _currentHyperlinkId if the custom id does not already exist
-        const auto result = _hyperlinkCustomIdMap.emplace(params, _currentHyperlinkId);
+        std::wstring newId{ id };
+        // hash the URL and add it to the custom ID - GH#7698
+        newId += L"%" + std::to_wstring(std::hash<std::wstring_view>()(uri));
```
Comment (DHowett) [inline src/buffer/out/textBuffer.cpp:2296]:
```suggestion
        newId += L"%" + std::to_wstring(std::hash<std::wstring_view>{}(uri));
```

prefer universal initialization syntax

**Review by miniksa — APPROVED** [review]

**Pull request merged**